container/container.go

```diff
@@ -47,151 +64,409 @@ var statuses = []struct {
-	StopService(namespace string) (err error)
-	DeleteVolume(name string) error
-}
-
-// DockerContainer provides high level interactions with Docker API for MESG.
-type DockerContainer struct {
+// Container starts and stops the MESG Service in Docker Container.
+type Container struct {
 	// client is a Docker client.
 	client client.CommonAPIClient
 
-	// namespace prefix.
-	nsprefix string
-
-	// sharedNetworkID is cache for an id network.
-	sharedNetworkID string
+	ipfsEndpoint   string
+	engineEndpoint string
+	engineName     string
+	engineNetwork  string
 }
 
-// New creates a new Container with given options.
-func New(nsprefix string) (*DockerContainer, error) {
+// New initializes the Container struct by connecting creating the Docker client.
+func New(ipfsEndpoint, engineName, engineAddress, engineNetwork string) (*Container, error) {
 	client, err := client.NewClientWithOpts(client.FromEnv)
 	if err != nil {
 		return nil, err
 	}
 	client.NegotiateAPIVersion(context.Background())
 
-	c := &DockerContainer{
-		client:   client,
-		nsprefix: nsprefix,
+	_, port, err := xnet.SplitHostPort(engineAddress)
+	if err != nil {
+		return nil, err
 	}
 
-	if err := c.isSwarmInit(); err != nil {
-		return nil, err
+	return &Container{
+		client:         client,
+		ipfsEndpoint:   ipfsEndpoint,
+		engineEndpoint: net.JoinHostPort(engineName, strconv.Itoa(port)),
+		engineName:     engineName,
+		engineNetwork:  engineNetwork,
+	}, nil
+}
+
+// Build the imge of the container
+func (c *Container) Build(srv *service.Service) error {
+	// download and untar service context into path.
+	path, err := ioutil.TempDir("", "mesg")
+	if err != nil {
+		return err
 	}
+	defer os.RemoveAll(path)
 
-	if err := c.createSharedNetwork(); err != nil {
-		return nil, err
+	resp, err := http.Get(c.ipfsEndpoint + srv.Source)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return errors.New("service's source code is not reachable, status: " + resp.Status + ", url: " + c.ipfsEndpoint + srv.Source)
+	}
+	defer resp.Body.Close()
+
+	if err := archive.Untar(resp.Body, path, &archive.TarOptions{ChownOpts: &idtools.Identity{
+		UID: os.Geteuid(),
+		GID: os.Getegid()},
+	}); err != nil {
+		return err
 	}
-	return c, nil
-}
 
-// Build builds a docker image.
-func (c *DockerContainer) Build(path string) (tag string, err error) {
 	excludeFiles, err := build.ReadDockerignore(path)
 	if err != nil {
-		return "", err
+		return err
 	}
 
 	tr, err := archive.TarWithOptions(path, &archive.TarOptions{
 		Compression:     archive.Gzip,
 		ExcludePatterns: excludeFiles,
 	})
 	if err != nil {
-		return "", err
+		return err
 	}
 	defer tr.Close()
 
-	resp, err := c.client.ImageBuild(context.Background(), tr, types.ImageBuildOptions{
+	if _, err := c.client.ImageBuild(context.Background(), tr, types.ImageBuildOptions{
 		Remove:         true,
 		ForceRemove:    true,
 		SuppressOutput: true,
-	})
+		Tags:           []string{imageTag + srv.Hash.String()},
```

Comment:
build with a tag 🙆‍♂

Reply:
You removed the `defer resp.Body.Close()`

Reply:
yes because we don't read it anymore and it doesn't seem to cause any issue.
should I put it back?

Reply:
Good point, if we don't read it I think it's fine